core/consensus/babe/src/epoch_changes.rs

```diff
@@ -0,0 +1,224 @@
+		parent_number: NumberFor<Block>,
+		slot_number: SlotNumber,
+		make_genesis: G,
+	) -> Result<Option<Epoch>, fork_tree::Error<D::Error>>
+		where G: FnOnce(SlotNumber) -> Epoch
+	{
+		use sr_primitives::traits::{One, Zero};
+
+		// find_node_where will give you the node in the fork-tree which is an ancestor
+		// of the `parent_hash` by default. if the last epoch was signalled at the parent_hash,
+		// then it won't be returned. we need to create a new fake chain head hash which
+		// "descends" from our parent-hash.
+		let fake_head_hash = fake_head_hash(parent_hash);
+
+		let is_descendent_of = descendent_of_builder
+			.build_is_descendent_of(Some((fake_head_hash, *parent_hash)));
+
+		if parent_number == Zero::zero() {
+			// need to insert the genesis epoch.
+			let genesis_epoch = make_genesis(slot_number);
+			let res = self.inner.import(
+				*parent_hash,
+				parent_number,
+				genesis_epoch.clone(),
+				&is_descendent_of,
+			);
+
+			match res {
+				Ok(_) | Err(fork_tree::Error::Duplicate) => {},
+				Err(e) => return Err(e),
+			}
+
+			return Ok(Some(genesis_epoch))
+		}
+
+
+		self.inner.find_node_where(
+			&fake_head_hash,
+			&(parent_number + One::one()),
+			&is_descendent_of,
+			&|epoch| epoch.start_slot <= slot_number,
+		)
+			.map(|n| n.map(|n| n.data.clone()))
+	}
+
+	/// Import a new epoch-change, signalled at the given block.
+	///
+	/// This assumes that the given block is prospective (i.e. has not been
+	/// imported yet), but its parent has. This is why the parent hash needs
+	/// to be provided.
+	pub fn import<D: IsDescendentOfBuilder<Block>>(
+		&mut self,
+		descendent_of_builder: D,
+		hash: Block::Hash,
+		number: NumberFor<Block>,
+		parent_hash: Block::Hash,
+		epoch: Epoch,
+	) -> Result<(), fork_tree::Error<D::Error>> {
+		let is_descendent_of = descendent_of_builder
+			.build_is_descendent_of(Some((hash, parent_hash)));
+
+		let res = self.inner.import(
+			hash,
+			number,
+			epoch,
+			&is_descendent_of,
+		);
+
+		match res {
+			Ok(_) | Err(fork_tree::Error::Duplicate) => Ok(()),
+			Err(e) => Err(e),
+		}
+	}
+}
+
+/// A shared epoch changes tree.
+#[derive(Clone)]
+pub struct SharedEpochChanges<Block: BlockT> {
+	inner: Arc<Mutex<EpochChanges<Block>>>,
+}
+
+impl<Block: BlockT> SharedEpochChanges<Block> {
+	/// Create a new instance of the `SharedEpochChanges`.
+	pub fn new() -> Self {
+		SharedEpochChanges {
+			inner: Arc::new(Mutex::new(EpochChanges::<Block>::new()))
+		}
+	}
+
+	/// Lock the shared epoch changes,
+	pub fn lock(&self) -> MutexGuard<EpochChanges<Block>> {
+		self.inner.lock()
+	}
+}
+
+impl<Block: BlockT> From<EpochChanges<Block>> for SharedEpochChanges<Block> {
+	fn from(epoch_changes: EpochChanges<Block>) -> Self {
+		SharedEpochChanges {
+			inner: Arc::new(Mutex::new(epoch_changes))
+		}
+	}
+}
+
+// TODO: write tests within the scope of this module.
+//
+// 1. mock up an `IsDescendentOfBuilder`. Don't use the `Client` - these tests should be
+// minimal and focused.
```

Comment:
https://github.com/paritytech/substrate/blob/master/core/utils/fork-tree/src/lib.rs#L649

We can mock an arbitrary tree like this.

Reply:
Would love to share that code!